dffml/source/file.py

```diff
@@ -49,10 +52,18 @@ def __repr__(self):
         elif self.filename[::-1].startswith(('.xz')[::-1]) or \
                 self.filename[::-1].startswith(('.lzma')[::-1]):
             opener = lzma.open(self.filename, 'rt')
+        elif zipfile.is_zipfile(self.filename):
+            archive = zipfile.ZipFile(self.filename)
+            for file in archive.infolist():
+                file = io.TextIOWrapper(archive.open(file, 'r'))
```

Comment:
TextIOWrapper inherits from https://docs.python.org/3/library/io.html#io.IOBase `io.IOBase` so we can use the `with` keyword here as well.

Hence we have the opportunity to keep the convention of using `opener = something`.

However, since we'll use the with keyword just on opener on line 65, we need to define a little helper function so that both the zipfile and the textiowrapper are closed at the end of the with block.

```python
from contextlib import contextmanager

...

elif zipfile.is_zipfile(self.filename):
    @contextmanager
    def opener_helper():
        with zipfile.ZipFile(self.filename) as archive:
            for file in archive.infolist():
                with io.TextIOWrapper(archive.open(filename, 'r')) as fd:
                    yield fd
            # Only care about the one file
            break
    opener = opener_helper()
```

Reply:
Done :+1: